starboard/nplb/maximum_player_configuration_explorer.h

```diff
@@ -0,0 +1,149 @@
+// considered less than or equal to the other vector if all the corresponding
+// integers in the first vector are smaller than or equal to the corresponding
+// integers in the second vector.
+// To identify the maximum elements, a test function, denoted as |test_functor|,
+// must be provided to evaluate whether the current configuration satisfies the
+// criterion for a maximum element.
+// The reason for applying the Depth First Search scheduler in this context is
+// that it is a faster method.
+class GradedPosetDFSScheduler {
+ public:
+  class HashFunction {
+   public:
+    size_t operator()(const std::vector<int>& v) const {
+      size_t r = 0;
+      for (const auto& it : v) {
+        r += it;
+        r *= 10;  // since we support at most 7 instances, 10 is good enough.
+      }
+      return r;
+    }
+  };
+
+  GradedPosetDFSScheduler(
+      int resources,
+      int instances,
+      const std::function<bool(const std::vector<int>&)>& test_functor);
+
+  std::set<std::vector<int>> GetMaxTestConfigSet() const {
+    return max_test_config_set_;
+  }
+
+ private:
+  void Run();
+
+  int resources_;
+  int instances_;
+  std::function<bool(const std::vector<int>&)> test_functor_;
+
+  std::set<std::vector<int>> max_test_config_set_;
+};
+
+// The |MaximumPlayerConfigurationExplorer| is employed to evaluate the maximum
+// feasible configurations for a group of players, with each player using a
+// distinct video/audio codec.
+// The search process is carried out using the |GradedPosetDFSScheduler|, while
+// the |Create| function is provided as the test function.
+class MaximumPlayerConfigurationExplorer {
+ public:
+  typedef std::tuple<SbMediaVideoCodec, SbMediaAudioCodec, SbPlayerOutputMode>
+      PlayerConfig;
+
+  static constexpr int kMaxPlayerInstance = 5;
+
+  MaximumPlayerConfigurationExplorer(
+      const std::set<PlayerConfig>& player_configs,
+      testing::FakeGraphicsContextProvider* fake_graphics_context_provider);
+
+  // The |pseudo_max_element_set| is utilized as the test patterns for
+  // identifying the maximum feasible configurations.
+  MaximumPlayerConfigurationExplorer(
+      const std::set<PlayerConfig>& player_configs,
+      const std::set<std::vector<int>>& pseudo_max_element_set,
+      testing::FakeGraphicsContextProvider* fake_graphics_context_provider);
+
+  std::set<std::vector<int>> GetMaxTestConfigSet() const {
+    return scheduler_->GetMaxTestConfigSet();
+  }
+
+  ~MaximumPlayerConfigurationExplorer();
+
+  MaximumPlayerConfigurationExplorer(
+      const MaximumPlayerConfigurationExplorer& other) = delete;
+  MaximumPlayerConfigurationExplorer& operator=(
+      const MaximumPlayerConfigurationExplorer& other) = delete;
+
+ private:
+  void Init();
+
+  // Test whether the system can support this particular configuration of
+  // players by creating it.
+  bool Create(const std::vector<int>& config);
+
+  std::unique_ptr<GradedPosetDFSScheduler> scheduler_;
```

Comment:
Move this after all member variables initialized in the ctor (especially if they will never be modified later).

Reply:
class GradedPosetDFSScheduler has been remvoed.